### ISSUE TITLE
Support Joinable loss functions with DDP

### ISSUE THREAD

**Issue by jbechara** [pytorch/pytorch#140107]
### 🚀 The feature, motivation and pitch

I'm working on a loss function that requires synchronizing a buffer across workers, and that **isn't** recoverable by a custom reduction of gradients across workers (which is already achievable through DDP's `register_comm_hook`).

I would like it to also work with the `Join` context manager for handling uneven dataloader lengths, i.e. make the loss function a `Joinable`.

As of today, DDP only has a single `JoinHook` that shadows any operations in **both its forward and backward passes**. Note that DDP does indeed perform some distributed operations in its forward pass, namely `_check_global_requires_backward_grad_sync` in `def _pre_forward`.

Since the loss function's forward must operate _after_ the model's forward pass and _before_ the model's backward pass, there is no ordering of joinables to `Join` (either `[ddp_model, loss]` or `[loss, ddp_model]`) that is able to produce correct behavior: the workers will always hang.

Potential resolutions from looking at the source code:
- separate DDP's join hooks into forward and backward join hooks that can allow the user to insert their `Joinable` in between, or,
- avoid any distributed operations in the forward, so that the current DDP's join hook can be assumed to only shadow the backward pass.

Please let me know if the use case is not clear, or if I can help by providing a concrete example.

### Alternatives

For now, my alternative is to split my loss function into two steps, where I synchronize the buffer in question prior to the ddp_model's forward call, and perform the loss function's forward call using the pre-computed buffer.

This happens to work in the case of my loss function, but is not necessarily doable for a loss function whose synchronized buffer depends on the output of the model's forward.

**Comment by fegin** [comment]
Can you provide an example? If the synchronization is done manually by the user, which is outside of DDP's scope, then extending DDP to support this feature seems not general. I may be misinterpret the issue, so an example would be good.
